syft/execution/plan.py

```diff
@@ -384,6 +450,48 @@ def __setattr__(self, name, value):
             for tensor_name, tensor in value.named_tensors():
                 self.__setattr__(f"{name}_{tensor_name}", tensor)
 
+    def validate_input(self, args, kwargs):
+        """
+        Function used in __call__ to validate the input based on the input received at build time.
+
+        Args:
+            args: original args received on __call__.
+            kwargs: original kwargs received on __call__.
+        """
```

Comment:
If I understand well, you're tagging the placeholder with a custom string to tell what is the "build time" type of this placeholder and then check that the runtime type is the same as what's written in the tag.
Wouldn't it be neater to add this information in the placeholder (as an attribute for instance) instead?

Reply:
It might be, you are right, I am not sure. I taught that this might not be wanted (to add a few new attributes), as there would be just a few placeholders that would actually use this.

Reply:
Another problem is that I still need a mechanism to flag nontensors inputs at build time, like lists for example, I considered that this would add some unwanted complexity to the Placeholder class.

Reply:
Maybe we can discuss about this with other people but I think it would be better to have this in the Placeholder class. This way, we could also have some kind of `validate_type` on instantiation and avoid the comparison of substrings (which I really don't like).

Reply:
@OpenMined/syft-core-team?

Reply:
> Another problem is that I still need a mechanism to flag nontensors inputs at build time, like lists for example, I considered that this would add some unwanted complexity to the Placeholder class.

I'll need to read what you do with non-tensor inputs in more detail but this should go in the Placeholder logic, you're right

Reply:
Yes. We want to make plans closer to how actual functions work (without them being actual functions). Let's say we want to build a plan with an underlying function that receives two parameters.

x: (and the build input) th.tensor([1,2])
y: (and the build input) [th.tensor([4,5]), th.tensor([6, 7])]

The generated placeholders:

placeholder for th.tensor([1, 2]) will have the following tags: input, fcall, type.
placeholder for [th.tensor([4, 5]), th.tensor([6, 7])] will have the following tags: fcall, type_list. (this is not an input tensor)
placeholder for th.tensor([4, 5]) will have the following tags: input
placeholder for th.tensor([6, 7]) will have the following tags: input
The fcall tag tells me that this placeholder is a placeholder for a build time argument (even if not FrameworkTensor). This is useful when doing the type check/arguments verification. In the old implementation, the list from the arguments would have not received a Placeholder, making type checking impossible.

The naming might not be the greatest, I'll document this better and think for a better name (suggestions would be great).

Reply:
Nontensor inputs would require a placeholders as well (for the typecheck), but if we create two attributes in Placeholder (something like known_type: str and build_argument: bool), the problem would be solved without tags.

Reply:
Yeah, let's preferentially use new attributes over string comparisons. 👍